### PR TITLE
Add toggle sneak setting

### PR DESCRIPTION
**Pull request by kotek900**
A compact, short information about this PR for easier understanding:

- Goal: Add toggle sneak option, when enabled toggles sneak instead of requiring it to be held down.
- #2194

## To do

This PR is a ~~Work in Progress~~ / Ready for Review.

- [x] Make code better (less ugly)
- [x] Add more info to the settings page

## How to test

Settings -> General -> Toggle sneak -> enable
Open a world and it should toggle sneak instead of requiring the button to be pressed continuously

### CODE REVIEW

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -119,6 +119,8 @@ repeat_place_time (Place repetition interval) float 0.25 0.16 2
 #    Automatically jump up single-node obstacles.
 autojump (Automatic jumping) bool false
 
+toggle_sneak (Toggle sneak) bool false
```
Comment (appgurueu) [inline builtin/settingtypes.txt:122]: An explanation (provided using `#`) would be good to have.

File: src/client/game.cpp [hunk 2]
```diff
@@ -709,6 +709,8 @@ struct GameRunData {
 	float time_from_last_punch;
 	ClientActiveObject *selected_object;
 
+	bool isSneaking = false;
```
Comment (appgurueu) [inline src/client/game.cpp:712]: `is_sneaking`

File: src/client/game.cpp [hunk 3]
```diff
@@ -2689,14 +2691,22 @@ void Game::updatePlayerControl(const CameraOrientation &cam)
 
 	//TimeTaker tt("update player control", NULL, PRECISION_NANO);
 
+
+	if (g_settings->getBool("toggle_sneak")) {
+		if (wasKeyPressed(KeyType::SNEAK)) runData.isSneaking = !runData.isSneaking;
```
Comment (appgurueu) [inline src/client/game.cpp:2696]: Put the `if` body on the next line to comply with our style guidelines.

File: src/client/game.cpp [hunk 4]
```diff
@@ -1445,7 +1447,7 @@ void Game::copyServerClientCache()
 {
 	// It would be possible to let the client directly read the media files
 	// from where the server knows they are. But aside from being more complicated
-	// it would also *not* fill the media cache and cause slower joining of 
+	// it would also *not* fill the media cache and cause slower joining of
```
Comment (appgurueu) [inline src/client/game.cpp:1450]: Not very important, but try to keep unrelated changes out of PRs

File: src/client/game.cpp [hunk 5]
```diff
@@ -2689,14 +2691,22 @@ void Game::updatePlayerControl(const CameraOrientation &cam)
...
+
 	PlayerControl control(
 		isKeyDown(KeyType::FORWARD),
 		isKeyDown(KeyType::BACKWARD),
 		isKeyDown(KeyType::LEFT),
 		isKeyDown(KeyType::RIGHT),
 		isKeyDown(KeyType::JUMP) || player->getAutojump(),
 		isKeyDown(KeyType::AUX1),
-		isKeyDown(KeyType::SNEAK),
+		runData.isSneaking,
```
Comment (appgurueu) [inline src/client/game.cpp:2709]: Not sure about this. Minetest is being inconsistent here. Controls are both "kind of" keybinds and "abstract" controls. The "forward" control for example does not take autoforward into account, while the "jump" control does take autojump into account.

I'm not sure whether we should take "sneak toggling" into account here. On the one hand, this is unsuitable for the use case of overloading actions using sneak (for example sneak + rightclick placing a node rather than opening a chest). On the other hand, how else can servers detect whether a player is presently sneaking?

Perhaps an option would be to separate this out into two fields (though the question of which of these to call `sneak` would persist)?

**Review by appgurueu — COMMENTED** [review]
A caveat this approach in general seems to have: It doesn't play well with the "sneak + control" overloading for special actions many mods do. You'd have to press sneak, then you'd be in "special action mode", and would have to press sneak to leave it again.

Arguably this is kind of the modder's fault, for tying sneaking to "special"/tertiary actions (but also kind of our fault, for not providing more aux. keybindings). Still it is a rather common pattern which is found in builtin and MTG.

File: src/client/game.cpp (hunk 5 above)
Comment (Zughy) [reply src/client/game.cpp]: @appgurueu any updates? Unsubscribing

**Comment by rubenwardy** [comment]
You also need to add `toggle_sneak` to here so that it appears in the accessibility tab: https://github.com/minetest/minetest/blob/master/builtin/mainmenu/settings/dlg_settings.lua#L83

File: src/client/game.cpp (hunk 5 above)
Comment (rubenwardy) [reply src/client/game.cpp]: I think that it's good as is - separating them makes the code more complicated and more confusing. You want to know whether the player is commanding a sneak, doesn't matter whether it's done by a key down, toggle, mouse button, or whatever

Comment (appgurueu) [reply src/client/game.cpp]: Can be left as-is.

**Comment by Zughy** [comment]
@kotek900 changes and rebase needed
